api: Fix narrowing conversion C++ compilation warning in Z_TIMEOUT_TICKS

r2r0 opened the pull request:
Assignement in Z_TIMEOUT_TICKS macro causes narrowing
conversion warning in Z_TIMEOUT_TICKS.
Explicit cast to type of ticks field makes compiler happy.

Fixes #26369

Signed-off-by: Artur Lipowski <Artur.Lipowski@hidglobal.com>

andyross commented:
Hm... I'm not sure we want to suppress that warning.  What this means is that user code is computing a ticks value in 64 bit precision and then passing it a kernel configured to use a 32 bit k_ticks_t type (c.f. CONFIG_TIMEOUT_64BIT).

If there's a need for a cast to promise that conversion is legal, then IMHO it should be in the application code, not the kernel.

Or am I missing something?

r2r0 commented:
> Hm... I'm not sure we want to suppress that warning. What this means is that user code is computing a ticks value in 64 bit precision and then passing it a kernel configured to use a 32 bit k_ticks_t type (c.f. CONFIG_TIMEOUT_64BIT).
...

This warning is emitted even if user just use k_msleep where there is no way to pass "too big" value.
IMHO the problem is that k_ms_to_ticks_ceil64 returns uint64_t and it is assigned to int64_t (type of ticks field).

andyross commented:
Then isn't that a bug in k_msleep(), that it's generating an inappropriately high precision value in a context where timeouts are smaller?  Again, the reason for using macros here is that we **want** to import the user's typing as it exists, and not throw data away.

r2r0 commented:
This warning will be generated for any use of Z_TIMEOUT_MS or Z_TIMEOUT_US or Z_TIMEOUT_NS or Z_TIMEOUT_CYC.
The value passed as parameter to above macros has no influence on this warning - compiler just warns about assigning unsigned integer value to signed integer (both the same bit length) can produce value narrowing.
Until ticks field is of type int64_t then it cannot accept the whole range of uint64_t (from k_ms_to_ticks_ceil64 nor from k_us_to_ticks_ceil64 etc.)

pabigot reviewed (changes requested):
I think it would be better to put the cast into the calls to `Z_TIMEOUT_TICKS()` as shown below.  The problem here is that the infrastructure has converted the user's input to a valid unsigned duration, but then passes it to a macro that allows for a negative duration.  The burden for clarification of intent is on the call site.  We should preserve the warning if somebody directly invokes `Z_TIMEOUT_TICKS()` with a value that would produce a narrowing internally.

```
diff --git a/include/sys_clock.h b/include/sys_clock.h
index 6af5447f42..938724472f 100644
--- a/include/sys_clock.h
+++ b/include/sys_clock.h
@@ -86,15 +86,15 @@ typedef struct {
 #define Z_FOREVER Z_TIMEOUT_TICKS(K_TICKS_FOREVER)
 
 #ifdef CONFIG_TIMEOUT_64BIT
-# define Z_TIMEOUT_MS(t) Z_TIMEOUT_TICKS(k_ms_to_ticks_ceil64(MAX(t, 0)))
-# define Z_TIMEOUT_US(t) Z_TIMEOUT_TICKS(k_us_to_ticks_ceil64(MAX(t, 0)))
-# define Z_TIMEOUT_NS(t) Z_TIMEOUT_TICKS(k_ns_to_ticks_ceil64(MAX(t, 0)))
-# define Z_TIMEOUT_CYC(t) Z_TIMEOUT_TICKS(k_cyc_to_ticks_ceil64(MAX(t, 0)))
+# define Z_TIMEOUT_MS(t) Z_TIMEOUT_TICKS((k_ticks_t)k_ms_to_ticks_ceil64(MAX(t, 0)))
+# define Z_TIMEOUT_US(t) Z_TIMEOUT_TICKS((k_ticks_t)k_us_to_ticks_ceil64(MAX(t, 0)))
+# define Z_TIMEOUT_NS(t) Z_TIMEOUT_TICKS((k_ticks_t)k_ns_to_ticks_ceil64(MAX(t, 0)))
+# define Z_TIMEOUT_CYC(t) Z_TIMEOUT_TICKS((k_ticks_t)k_cyc_to_ticks_ceil64(MAX(t, 0)))
 #else
-# define Z_TIMEOUT_MS(t) Z_TIMEOUT_TICKS(k_ms_to_ticks_ceil32(MAX(t, 0)))
-# define Z_TIMEOUT_US(t) Z_TIMEOUT_TICKS(k_us_to_ticks_ceil32(MAX(t, 0)))
-# define Z_TIMEOUT_NS(t) Z_TIMEOUT_TICKS(k_ns_to_ticks_ceil32(MAX(t, 0)))
-# define Z_TIMEOUT_CYC(t) Z_TIMEOUT_TICKS(k_cyc_to_ticks_ceil32(MAX(t, 0)))
+# define Z_TIMEOUT_MS(t) Z_TIMEOUT_TICKS((k_ticks_t)k_ms_to_ticks_ceil32(MAX(t, 0)))
+# define Z_TIMEOUT_US(t) Z_TIMEOUT_TICKS((k_ticks_t)k_us_to_ticks_ceil32(MAX(t, 0)))
+# define Z_TIMEOUT_NS(t) Z_TIMEOUT_TICKS((k_ticks_t)k_ns_to_ticks_ceil32(MAX(t, 0)))
+# define Z_TIMEOUT_CYC(t) Z_TIMEOUT_TICKS((k_ticks_t)k_cyc_to_ticks_ceil32(MAX(t, 0)))
 #endif
 
 /* Converts between absolute timeout expiration values (packed into
```

r2r0 committed:
api: Fix narrowing conversion C++ compilation warning in Z_TIMEOUT_TICKS

Assignement in Z_TIMEOUT_TICKS macro causes narrowing
conversion warning in Z_TIMEOUT_TICKS.
Explicit cast to type of ticks field makes compiler happy.

Fixes #26369

Signed-off-by: Artur Lipowski <Artur.Lipowski@hidglobal.com>

andyross reviewed (approved):
This seems more reasonable as it touches only internal APIs and leaves the warning in place for user code doing truncation (I didn't realize this was purely a sign warning earlier).  Thanks.
